### PR TITLE
BUG: Categorical(Index) passed as categories

### LINKED ISSUE #17884: Categorical.from_codes does not allow categories to be a CategoricalIndex
#### Code Sample, a copy-pastable example if possible

`Categorical.from_codes` requires `categories` to be specified, with documented type "index-like" but passing a `CategoricalIndex` raises an error.

```python
>>> pd.Categorical.from_codes([1, 2], pd.CategoricalIndex(['Hello', 'world', '!', '**'])
/Users/joel/anaconda3/envs/scipy3k/lib/python3.5/site-packages/pandas/core/categorical.py in get_values(self)
   1276         if is_datetimelike(self.categories):
   1277             return self.categories.take(self._codes, fill_value=np.nan)
-> 1278         return np.array(self)
   1279
   1280     def check_for_ordered(self, op):

ValueError: object __array__ method not producing an array
```

#### Problem description

It seems that being able to construct a categorical column with an existing `CategoricalIndex` would be a useful way to share `CategoricalIndex` int-string mappings. This would seem to be what `from_codes` aims to facilitate, but requires reconstructing the `CategoricalIndex`, which seems unnecessary to me.

The documentation also says index-like which seems underspecified if a CategoricalIndex is not accepted. At least, the error is not sufficiently informative.

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
- [x] closes #17884
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

@TomAugspurger I am not sure we had a discussion about that before, but to me it does not make many sense to have a CategoricalIndex as the categories of a Categorical/CategoricalIndex, so I think we should convert passed categories (which is what this PR is doing). 
Alternative to fix #17884 is to fix the repr code to deal with Categorical categories.

### CODE REVIEW

File: pandas/core/dtypes/dtypes.py [hunk 1]
```diff
@@ -316,7 +315,10 @@ def _validate_categories(categories, fastpath=False):
         from pandas import Index
 
         if not isinstance(categories, ABCIndexClass):
-            categories = Index(categories)
+            categories = Index(categories, tupleize_cols=False)
+
```
Comment (jreback) [inline pandas/core/dtypes/dtypes.py]: should be an elif and instead should use is_categorical

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/dtypes/dtypes.py (hunk 1 above)
Comment (jorisvandenbossche) [reply pandas/core/dtypes/dtypes.py]: Isn't the point of the ABC ones to avoid imports from itself ? (as `is_categorical` uses CategoricalDtype)

**Comment by TomAugspurger** [comment]
Thanks Joris, agreed with the changes here. Looks like this conflicts with your fastpath PR.

File: pandas/core/dtypes/dtypes.py (hunk 1 above)
Comment (jorisvandenbossche) [reply pandas/core/dtypes/dtypes.py]: @jreback Can you answer this one? No problem to change it if you think it is better, but it seems to me using `isinstance(categories, ABCCategoricalIndex)` is consistent with the use of `isinstance(categories, ABCIndexClass)` just above (also in readability of the code)

File: pandas/tests/test_categorical.py [hunk 2]
```diff
@@ -519,6 +519,18 @@ def test_contructor_from_categorical_string(self):
...
+        # GH17884
+        expected = pd.Categorical(['a', 'b'], categories=['a', 'b', 'c'])
+
+        result = pd.Categorical(
+            ['a', 'b'], categories=pd.Categorical(['a', 'b', 'c']))
+        tm.assert_categorical_equal(result, expected)
+
+        result = pd.Categorical(
+            ['a', 'b'], categories=pd.CategoricalIndex(['a', 'b', 'c']))
+        tm.assert_categorical_equal(result, expected)
```
Comment (jschendel) [inline pandas/tests/test_categorical.py]: `Categorical` and `CategoricalIndex` have been imported, so you can remove the `pd.` here, if you feel that's cleaner.

File: pandas/tests/test_categorical.py [hunk 3]
```diff
@@ -560,6 +572,18 @@ def f():
...
+        # GH17884
+        expected = pd.Categorical(['a', 'b'], categories=['a', 'b', 'c'])
+
+        result = pd.Categorical.from_codes(
+            [0, 1], categories=pd.Categorical(['a', 'b', 'c']))
+        tm.assert_categorical_equal(result, expected)
+
+        result = pd.Categorical.from_codes(
+            [0, 1], categories=pd.CategoricalIndex(['a', 'b', 'c']))
+        tm.assert_categorical_equal(result, expected)
```
Comment (jschendel) [inline pandas/tests/test_categorical.py]: same regarding `pd.`

File: pandas/core/dtypes/dtypes.py (hunk 1 above)
Comment (jreback) [reply pandas/core/dtypes/dtypes.py]: hmm, yeah I guess you have to do this. but you can also check ``isinstance(categories, (ABCCateoricalIndex, ABCCategorical))`` is more generic.

File: doc/source/whatsnew/v0.21.0.txt [hunk 4]
```diff
@@ -1023,6 +1023,7 @@ Categorical
 - Bug in the categorical constructor with empty values and categories causing the ``.categories`` to be an empty ``Float64Index`` rather than an empty ``Index`` with object dtype (:issue:`17248`)
 - Bug in categorical operations with :ref:`Series.cat <categorical.cat>` not preserving the original Series' name (:issue:`17509`)
 - Bug in :func:`DataFrame.merge` failing for categorical columns with boolean/int data types (:issue:`17187`)
+- Bug in constructing a ``Categorical``/``CategoricalDtype`` when the specified ``categories`` where of categorical type (:issue:`17884`).
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: I would move this to the CDT issue (as its actually not an independent bug) and better there.

Comment (jorisvandenbossche) [reply doc/source/whatsnew/v0.21.0.txt]: The bug also exists for `Categorical` and `CateforicalIndex` and was already present in current released 0.20.3. So not directly related to the CDT rework (although the bug is present in CDT as well). But not that important, can also move if you want.

File: pandas/tests/test_categorical.py (hunk 2 above)
Comment (jorisvandenbossche) [reply pandas/tests/test_categorical.py]: yes, thanks! (updated)

File: pandas/core/dtypes/dtypes.py (hunk 1 above)
Comment (jorisvandenbossche) [reply pandas/core/dtypes/dtypes.py]: updated to use `isinstance(categories, (ABCCateoricalIndex, ABCCategorical))` first and then only to check for Index

**Comment by jorisvandenbossche** [comment]
One disadvantage of the my changes is that is somehow relaxes the "unique" requirement for the passed categories. Because now you can pass a non-unique Categorical, and we don't raise anymore but just take its unique categories. 
I think it seems obvious that this is what the user want, but still, I am not sure relaxing the requirement is a good idea (eg in `from_codes` it means the passed codes don't match exactly the passed categories). But actually I can easily check this by moving the check for Categorical values after the check for unique / nulls.

**Comment by jreback** [comment]
@jorisvandenbossche ping when updated.

**Comment by jorisvandenbossche** [comment]
As far as I am concerned, this is updated. Is there a comment I overlooked?

File: pandas/core/dtypes/dtypes.py [hunk 5]
```diff
@@ -326,6 +326,9 @@ def _validate_categories(categories, fastpath=False):
             if not categories.is_unique:
                 raise ValueError('Categorical categories must be unique')
 
+        if isinstance(categories, ABCCategoricalIndex):
```
Comment (jreback) [inline pandas/core/dtypes/dtypes.py]: here you an also accept ABCCategorical

Comment (jorisvandenbossche) [reply pandas/core/dtypes/dtypes.py]: a Categorical has already been converted to CategoricalIndex

Comment (jreback) [reply pandas/core/dtypes/dtypes.py]: ahh ok

**Review by jreback — APPROVED** [review]

**Pull request merged**